### PR TITLE
get all users end point - validation and testing

### PR DESCRIPTION
**Pull request by PedrOrtizM**
## Summary

- GET: /users end point 
- Session validation 
- Unit test

## Trello Card
https://trello.com/c/NFvdlTfj/3-lista-de-usuarios

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -35,3 +35,15 @@ exports.signIn = async (req, res, next) => {
     next(error);
   }
 };
+
+exports.getUsers = async (req, res, next) => {
+  try {
+    const { limit, page } = req.query;
+    const offset = page * limit;
```
Comment (dbenavidesv) [inline app/controllers/user.js:42]: This should be like this, so the offset can be 0
```suggestion
    const offset = (page - 1) * limit;
```

File: app/helpers/schemas/paginationSchema.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  properties: {
+    limit: {
+      type:'string',
+      pattern: '^[0-9]*$',
+      default: '10'
+    },
+    page: {
+      type:'string',
+      pattern: '^[0-9]*$',
+      default: '0'
```
Comment (dbenavidesv) [inline app/helpers/schemas/paginationSchema.js:12]: Normally page number starts at 1

File: app/middlewares/validateSchema.js [hunk 3]
```diff
@@ -2,12 +2,12 @@ const Ajv = require('ajv');
...
 
-const ajv = new Ajv();
+const ajv = new Ajv({ useDefaults: true });
 addFormats(ajv);
 
-module.exports.validateSchema = schema => (req, res, next) => {
+module.exports.validateSchema = (schema, type = 'body') => (req, res, next) => {
   const validate = ajv.compile(schema);
-  if (validate(req.body)) return next();
+  if (validate(req[type])) return next();
```
Comment (dbenavidesv) [inline app/middlewares/validateSchema.js:10]: nice

File: app/middlewares/validateSession.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const { TOKEN_ERROR } = require("../constants/errors");
+const { authenticationError, authorizationError } = require("../errors");
+const { verifyToken } = require("../helpers/jwt");
+const logger = require('../logger');
+
+module.exports.validateSession = async (req, res, next) => {
+
+  const token = req.headers.authorization;
+  if (!token)
+    next(authenticationError(TOKEN_ERROR));
```
Comment (dbenavidesv) [inline app/middlewares/validateSession.js:10]: return here

File: app/routes.js [hunk 5]
```diff
@@ -1,10 +1,13 @@
...
-const { signUp, signIn } = require('./controllers/user');
+const { signUp, signIn, getUsers } = require('./controllers/user');
 const { validateSchema } = require('./middlewares/validateSchema');
 const { userSchema, credentialsSchema } = require('./helpers/schemas/userSchema');
+const { paginationSchema } = require('./helpers/schemas/paginationSchema');
+const { validateSession } = require('./middlewares/validateSession');
 
 exports.init = app => {
   app.get('/health', healthCheck);
+  app.get('/users',  [validateSession,validateSchema(paginationSchema, 'query')], getUsers);
```
Comment (dbenavidesv) [inline app/routes.js:10]: create constants for the  `query`, `body`, `params` keywords

Comment (dbenavidesv) [inline app/routes.js:10]:
```suggestion
  app.get('/users',  [validateSession, validateSchema(paginationSchema, 'query')], getUsers);
```

File: app/helpers/schemas/paginationSchema.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+module.exports.paginationSchema = {
+  type: 'object',
+  properties: {
+    limit: {
+      type:'string',
```
Comment (dbenavidesv) [inline app/helpers/schemas/paginationSchema.js:5]: shouldn't this be a number type?

File: app/helpers/schemas/paginationSchema.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+      pattern: '^[0-9]*$',
+      default: '10'
+    },
+    page: {
+      type:'string',
```
Comment (dbenavidesv) [inline app/helpers/schemas/paginationSchema.js:10]: same here

File: test/factory/user.js [hunk 8]
```diff
@@ -5,3 +5,4 @@ const { factoryWithCustomizedValue } = require('./factory_by_models');
 
 factoryWithCustomizedValue('User', 'password', encrypt(credentialsMock.password));
 exports.factoryUser = attributes => factory.create('User', attributes);
+exports.factoryManyUser = (number,attributes) => factory.createMany('User',number,attributes);
```
Comment (dbenavidesv) [inline test/factory/user.js:8]: maybe `create` would be a better word for these two names

Comment (dbenavidesv) [inline test/factory/user.js:8]:
```suggestion
exports.createUser = attributes => factory.create('User', attributes);
exports.createManyUsers = (number, attributes) => factory.createMany('User', number, attributes);
```

File: app/middlewares/errors.js [hunk 9]
```diff
@@ -7,6 +7,7 @@ const statusCodes = {
   [errors.DATABASE_ERROR]: 503,
   [errors.DEFAULT_ERROR]: 500,
   [errors.AUTHENTICATION_ERROR]: 401,
+  [errors.AUTHORIZATION_ERROR]: 403,
```
Comment (dbenavidesv) [inline app/middlewares/errors.js:10]: I don't think 403 is the appropriate code for this, you should use `401` as well check this, related to `403`:
```The request was a legal request, but the server is refusing to respond to it. Unlike a 401 Unauthorized response, authenticating will make no difference.```

**Review by dbenavidesv — COMMENTED** [review]
One more thing, execute a `npm run lint-fix` before pushing

File: app/helpers/schemas/paginationSchema.js (hunk 6 above)
Comment (PedrOrtizM) [reply app/helpers/schemas/paginationSchema.js:5]: Yes, but the query params are strings. When AJV tries to validate it returns an error.

> with type: 'number' 

image

Comment (dbenavidesv) [reply app/helpers/schemas/paginationSchema.js:5]: Got it. Can you check if it's a numeric string? the same with the page parameter

Comment (dbenavidesv) [reply app/helpers/schemas/paginationSchema.js:5]: Ah, I just saw that you did just that with a regex, is there a way to do it with an AJV option? just for knowledge haha

**Review by dbenavidesv — APPROVED** [review]

File: app/helpers/schemas/paginationSchema.js (hunk 6 above)
Comment (PedrOrtizM) [reply app/helpers/schemas/paginationSchema.js:5]: Yes, another solution is to enable Type Coercion when creating the AJV instance.
`const ajv = new Ajv({coerceTypes: true})`
in that case, you can use `type: number` in schema validator

**Pull request merged**